Update for latest CSI driver image post issues/167

2uasimojo opened the pull request:
Since the [fix](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/185) for [issue #167](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/167) merged upstream, the AWS EFS CSI driver team overwrote the [`latest` image tag](sha256-962619a5deb34e1c4257f2120dd941ab026fc96adde003e27f70b65023af5a07?context=explore) to include it.

For starters, this means we can update this operator to use the [new method of specifying access points](https://github.com/kubernetes-sigs/aws-efs-csi-driver/tree/0ae998c5a95fe6dbee7f43c182997e64872695e6/examples/kubernetes/access_points#edit-persistent-volume-spec) via a colon-delimited `volumeHandle` as opposed to in `mountOptions`.

However, the same update to `latest` also brought in a [commit](https://github.com/kubernetes-sigs/aws-efs-csi-driver/commit/b3baff821a477161599eef757c5ed59dc7126073) that requires an additional mount in the `efs-plugin` container of the DaemonSet. So we need to update our YAML for that resource at the same time, or everything is broken (this might be upstream [issue #192](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/192)). This update to the DaemonSet YAML also syncs with [upstream](https://github.com/kubernetes-sigs/aws-efs-csi-driver/blob/0ae998c5a95fe6dbee7f43c182997e64872695e6/deploy/kubernetes/base/node.yaml) by bumping the image versions for the other two containers (csi-node-driver-registrar: v1.1.0 => v1.3.0; and livenessprobe: v1.1.0 => livenessprobe:v2.0.0).

Jira: [OSD-4187](https://issues.redhat.com/browse/OSD-4187)

2uasimojo committed:
Update for latest CSI driver image post issues/167

Since the [fix](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/185) for [issue #167](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/167) merged, the AWS EFS CSI driver overwrote the [`latest` image tag](sha256-962619a5deb34e1c4257f2120dd941ab026fc96adde003e27f70b65023af5a07?context=explore) to include it.

For starters, this means we can update this operator to use the [new method of specifying access points](https://github.com/kubernetes-sigs/aws-efs-csi-driver/tree/0ae998c5a95fe6dbee7f43c182997e64872695e6/examples/kubernetes/access_points#edit-persistent-volume-spec) via a colon-delimited `volumeHandle` as opposed to in `mountOptions`.

However, the same update to `latest` also brought in a [commit](https://github.com/kubernetes-sigs/aws-efs-csi-driver/commit/b3baff821a477161599eef757c5ed59dc7126073) that requires an additional mount in the `efs-plugin` container of the DaemonSet. So we need to update our YAML for that resource at the same time, or everything is broken (this might be upstream [issue #192](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/192). This update to the DaemonSet YAML also syncs with [upstream](https://github.com/kubernetes-sigs/aws-efs-csi-driver/blob/0ae998c5a95fe6dbee7f43c182997e64872695e6/deploy/kubernetes/base/node.yaml) by bumping the image versions for the other two containers (csi-node-driver-registrar: v1.1.0 => v1.3.0; and livenessprobe: v1.1.0 => livenessprobe:v2.0.0).

2uasimojo commented on pkg/controller/sharedvolume/sharedvolume_controller.go:
```diff
@@ -170,6 +170,12 @@ func (r *ReconcileSharedVolume) Reconcile(request reconcile.Request) (reconcile.
 	pve := pvEnsurable(sharedVolume)
 	pvce := pvcEnsurable(sharedVolume)
 
+	// TODO: If we just upgraded past 0.0.2, this will update the PV from old style (access point
+	// in MountOptions) to new style (access point in VolumeHandle). Which would be fine, except
+	// - I think the Update will bounce because PVs are immutable after creation. That could get
+	//   us in a hard loop.
+	// - If it doesn't, we've still got https://github.com/openshift/origin/issues/2434 which may
+	//   cause the update to wedge.
```
We may want to address this by simply never trying to update a PV. If I can confirm that it won't work anyway, I'll make that change. (The easiest way would be to use `alwaysEqual` -- with explanatory comments! -- for PVs (and conceivably PVCs as well?).)

2uasimojo replied:
This is done in https://github.com/openshift/aws-efs-operator/pull/17/commits/ec68e27ec44e8ef79a61ac005bff00bac71aa9d7

2uasimojo committed:
Refactor: Move alwaysEqual and equalOtherThanMeta

Make these methods exported from the util package, next to the Ensurable
interface, since a) a subsequent commit is going to use the former from
the sharedvolume side, and b) they make sense there anyway, since
there's nothing statics-specific about them.

2uasimojo committed:
Don't try to update PVs

PVs are immutable once created. This *should* mean we can never hit a
code path where we would *need* to `Update` a PV (via `Ensure`). And if
we did, it would bounce anyway because... PVs are immutable once
created.

There's a weird edge case, though: if the shape of a PV created by this
operator changes from one version to the next, the latter will come up
and think the PV needs to be changed to the new shape, and will try to
`Update`. That would bounce and put the operator into a hot loop.

So this commit simply disables `Update`s of PVs by pretending they never
change. We'll still restore one if deleted (which probably also won't
work, for reasons).

2uasimojo commented:
I live tested this with a couple of SV/PV/PVC groups from the previous version, and they are left alone as expected. No hot looping in the operator. Deleting the old SVs deletes their respective PV/PVC pairs without complaint. This is ready to go.

jharrington22 commented on pkg/controller/sharedvolume/ensurables_test.go:
```diff
@@ -130,12 +132,14 @@ func TestCache(t *testing.T) {
 	// PVs
 	pv1 := pvEnsurable(&sharedVolume).(*util.EnsurableImpl).Definition.(*corev1.PersistentVolume)
-	if pv1.Spec.CSI.VolumeHandle != fakeFSID {
+	expVolumeHandle1 := fakeFSID + "::" + fakeAPID
```
Consistent and use `fmt.Sprintf()` everywhere? I also find it easier to read.

jharrington22 commented on pkg/controller/sharedvolume/pv_ensurable.go:
```diff
@@ -28,7 +24,13 @@ func pvEnsurable(sharedVolume *awsefsv1alpha1.SharedVolume) util.Ensurable {
+			// The exception is an upgrade like the one from 0.0.2, where the shape of a SV-backed
+			// PV changed, meaning that if the operator notices an old-style PV, it will try to
+			// "fix" it. Which won't work. Spoofing "always equal" will avoid that.
+			EqualFunc: util.AlwaysEqual,
```
Instead of having this always be `true` can we just not `watch` PV's?

jharrington22 commented on pkg/controller/sharedvolume/sharedvolume_controller.go:
```diff
@@ -170,6 +170,12 @@ func (r *ReconcileSharedVolume) Reconcile(request reconcile.Request) (reconcile.
 	pve := pvEnsurable(sharedVolume)
 	pvce := pvcEnsurable(sharedVolume)
 
+	// TODO: If we just upgraded past 0.0.2, this will update the PV from old style (access point
```
Assuming you are talking about OperatorHub 0.0.2 here? We might want a little more context if so.

jharrington22 commented on pkg/controller/sharedvolume/sharedvolume_controller_test.go:
```diff
@@ -304,36 +303,88 @@ func TestReconcile(t *testing.T) {
+	if err = r.client.Delete(ctx, pvMap[pvname]); err != nil {
+		t.Fatal(err)
+	}
+	if res, err = r.Reconcile(req); res != test.NullResult || err != nil {
```
A `Delete` should cause a requeue no?

2uasimojo replied:
No, but I'm not sure which of the following you were thinking of when you asked:
1) There's no controller running here, so no watches are triggering, and no events are happening automatically. The test case is simulating the controller, manually firing individual Reconciles. In this case, we're pretending to get triggered by the deletion of the PV/PVC. (IRL those could come in as separate events, or if they happen "together", they'll be collapsed because they both get resolved to the same SV. I didn't bother testing the former case.)
2) The Reconcile restores the PV and PVC (via `Ensure`) the first time through, so it doesn't need to return a requeue.

jharrington22 commented on pkg/controller/sharedvolume/sharedvolume_controller_test.go:
```diff
@@ -343,11 +394,31 @@ func TestReconcile(t *testing.T) {
+	if err = r.client.Update(ctx, pv); err != nil {
+		t.Fatal(err)
+	}
+	if res, err = r.Reconcile(req); res != test.NullResult || err != nil {
```
Same here with an `Update` you should see a requeue

2uasimojo replied:
1) See [above](https://github.com/openshift/aws-efs-operator/pull/17/files#r445221682).
2) In this case Reconcile sees the malformed PV, but can't do anything about it, so it logs but "succeeds". (This is where we might have considered deleting and recreating the PV/PVC pair, but that wouldn't work with pods attached, etc. Besides, this should never happen anyway because PVs are immutable once created.)

2uasimojo committed:
Use `fmt.Sprintf` instead of `+`

Per [review](https://github.com/openshift/aws-efs-operator/pull/17/files#r445217067), use `fmt.Sprintf` rather than `+` to construct strings.

2uasimojo committed:
Fix references to 0.0.2

Some comments referred to "0.0.2" as the version before we changed the shape of PVs to use the colon-delimited VolumeHandle. But 0.0.2 refers to the [community-operators](https://github.com/operator-framework/community-operators/tree/master/community-operators/aws-efs-operator/0.0.2) (OperatorHub) version, which is just one of many possible version numbers that could refer to the same thing.

Clarify these by instead referring to the commit where the change happened. Let the reader figure out what version that corresponds to in whatever context they care about.

2uasimojo reviewed (commented):
Thanks James. Two commits on the way.

jharrington22 reviewed (approved):
/lgtm

jharrington22 commented:
/hold

jharrington22 commented:
/hold cancel

2uasimojo commented:
Grr, why isn't PR check completing?

2uasimojo commented:
/retest

2uasimojo commented:
/retest
